Tools/cases_generator/generate_cases.py

```diff
@@ -388,7 +425,13 @@ def write_body(self, out: Formatter, dedent: int, cache_adjust: int = 0) -> None
         assert dedent <= 0
         extra = " " * -dedent
         names_to_skip = self.unmoved_names | frozenset({UNUSED, "null"})
+        offset = 0
+        context = self.block.context
+        assert context != None
+        filename = context.owner.filename
         for line in self.block_text:
+            out.set_lineno(self.block_line + offset, filename)
```

Comment:
If you passed tokens, instead of strings, to the Formatter, it could handle the line numbers internally.